Sending StatsD events for non-fatal errors

dterror-zz opened the pull request:
File errors and ETL errors that get rescued by the system get logged at the end of the execution. We want to also emit an event to StatsD so we properly track it.


Mmm, so I was able to test it locally but reproducing a kafka reading exception.

@olessia `camus-errors` too generic a metric name?

dterror-zz committed:
Sending StatsD events for non-fatal errors

File errors and ETL errors that get rescued by the system get logged at
the end of the execution. We want to also emit an event to StatsD so we
properly track it.

olessia commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/CamusJob.java:
```diff
@@ -534,12 +535,14 @@ private boolean checkExecutionErrors(final FileSystem fs,
       final List<Pair<EtlKey, ExceptionWritable>> errorsFromFile = fileEntry.getValue();
       if (errorsFromFile.size() > 0) {
         log.error("Errors from file [" + filePath + "]");
+        StatsdReporter.gauge(configuration, "camus-errors", 1L);
```
I would not report this here. If we only report per ETL key that maps nicely to partitions, so we'll know how many partitions had trouble by simply doing the sum.

dterror replied:
makes sense, I'll keep etl only

olessia commented:
`camus-errors` is generic _enough_, since we don't actually check what the error was. Now, just heads up, this would also catch things like the "max pull time" which we already alert on and which is not fatal. Might be a bit chatty, but no harm in trying it out

dterror commented:
`max pull time` dang that's true. It will get chatty, I wonder if I can filter those (I'll try it out).

dterror-zz committed:
only reporting errors per EtlKey, not per file

dterror-zz committed:
Adding statsd logging also for partitions too close to earliest offset

Currently Camus sends an email for those, adding statsd so we can better
track those events.

dterror commented:
I've added instrumentation for `close-to-earliest-offset` as well.

olessia commented:
That's a great idea, but maybe we can change the alert condition? The email is very chatty, and ideally we would filter things like `The current offset is too close to the earliest offset, Camus might be falling behind: trekkie.overseer.clickstream uri:tcp://kafka1.data-chi.shopifydc.com:9092	leader:1	partition:1	earliest_offset:187	offset:187	latest_offset:189 avg_msg_size:2906	estimated_size:5812`.  What do you think?

dterror-zz committed:
Setting a min threshold on the absolute number of offsets to processs

Before we alert that the topic might be falling behind

dterror commented:
I've tried to reduce the noise by adding an absolute threshold on the check for `offset too close to earliest offset` log/email. It's not perfect, but it'll exclude cases where the processing is not falling behind, but the topic is very sparse.

dterror commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java:
```diff
@@ -86,6 +87,11 @@
+  // FALLING_BEHIND_ALERT_OFFSETS_MULTIPLER times the number-of-offsets-left-in-retention ....
+  public static final int FALLING_BEHIND_ALERT_OFFSETS_MULTIPLER = 3;
+  /// ...but only if the absolute number of offsets to process is greater than FALLING_BEHIND_ALERT_OFFSETS_THRESHOLD
+  public static final int FALLING_BEHIND_ALERT_OFFSETS_THRESHOLD = 1000000;
```
too much?

olessia replied:
I think the reason the original logic uses a multiplier is because an absolute threshold is topic-dependent. With too small a number it's too chatty, with too large -- it will miss a legitimate case of partition being stuck. I've actually figured out how to report on stuck offsets from Splunk, _I think_. Maybe the best solution in the meantime is to not mess with the condition but just to report to datadog the same thing that was reported via email before (just so we have alerts in one place) and we can periodically (e.g. twice a day) check up on those alerts until we convince ourselves that the spunk alert works?

dterror replied:
Adding the absolute check in _addition to_ the multiplier would just remove most of the noise, it's pragmatic, I think. But I agree that once we have the splunk alerts, we'll be able to be less stringent on this particular data point. So, I'll remove this and rely on the splunk alerts for stuck partitions.

dterror-zz committed:
Removing the absolute volume check, keeping the original multiplier

But maintaining the refactor which makes it easier to read the code
